### PR TITLE
Feature: provide automated PR reviews for linting related changes

### PR DESCRIPTION
**Pull request by OnkarRuikar**
- Alternative to https://github.com/mdn/content/pull/28228

Purpose of this PR is to provide an alternative approach. This PR involves all the linters/formatters (markdown, prettier, front-matter) for automated review using [reviewdog](https://github.com/reviewdog/reviewdog).

## Summary

Current Prettier implementation in our CI is being annoying because the formatter tells contributors that there is something wrong with their changes without telling what is it and how to fix it. This is forcing contributors to fix those issues using command line tools or online Prettier playground.\
Using reviewdog it is possible to instead suggest those changes as review comments on PRs. With this contributors will simply have to accept these changes and commit them in a batch.

While we are at it we can involve markdownlint and front-matter linters too, i.e we can suggest solutions to markdown and front-matter related issues as well.

## Implementation

reviewdog uses two inputs to provide review comments:
a) `git diff` generated after running the linters with `--fix` or `--write` options
b) logs generated by the linters along with error message matching pattern

The implementation performs following steps to provide the review comments:
1) Run all the linters/formatter to fix the code and collect logs along with exit statuses.
2) Use `git diff` to suggest changes done by the linters using `reviewdog -f=diff` command. Basically we are suggesting the same changes done by the linters.
**Note:** The diff is collected after running all the linters because we always run Prettier the last in order to avoid clashes with markdownlint. Also providing reviews for each linter separately will confuse the contributors. The combined linting is being presented as "mdn-linter".
3) If there are unfixable issues reported by markdownlint then put them as review comments using `reviewdog -efm` command.
4) Initially the front-matter linter had been implemented as a yari tool, so the logs generated by it are multi line. Till we make those logs single line just log the front-matter output and fail the workflow.

### CODE REVIEW

File: .github/workflows/pr-review-lint.yml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      - name: Lint and format markdown files
+        run: |
+          #echo "::add-matcher::.github/workflows/markdownlint-problem-matcher.json"
+          echo "Running markdownlint --fix"
+          files_to_lint="${{ env.DIFF_DOCUMENTS }}"
+          MD_LINT_STATUS=0
+          MD_LINT_LOG=$(yarn markdownlint-cli2 --fix ${files_to_lint} 2>&1) || MD_LINT_STATUS=1
+          echo "MD_LINT_LOG<<EOF" >> $GITHUB_ENV
+          echo "${MD_LINT_LOG}" >> $GITHUB_ENV
+          echo "EOF" >> $GITHUB_ENV
```
Comment (yin1999) [inline .github/workflows/pr-review-lint.yml:55]: We could also use openssl to generate a random string as a delimiter.

https://github.com/yin1999/get-changed-files/blob/49bb776318e6a1edffbc8b27725e49a637e9efec/action.yml#L28-L31

File: .github/workflows/pr-review-lint.yml [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+          # https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#multiline-strings
+          EOF="$(openssl rand -hex 8)"
+
+          echo "Running markdownlint --fix"
+          files_to_lint="${{ env.DIFF_DOCUMENTS }}"
+          MD_LINT_STATUS=0
+          MD_LINT_LOG=$(yarn markdownlint-cli2 --fix ${files_to_lint} 2>&1) || MD_LINT_STATUS=1
+          echo "MD_LINT_LOG<<${EOF}" >> $GITHUB_ENV
+          echo "${MD_LINT_LOG}" >> $GITHUB_ENV
+          echo "EOF" >> $GITHUB_ENV
```
Comment (yin1999) [inline .github/workflows/pr-review-lint.yml:58]:
```suggestion
          echo "${EOF}" >> $GITHUB_ENV
```

**Review by yin1999 — COMMENTED** [review]
One nit :)

File: .github/workflows/pr-review-lint.yml [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+          echo "${MD_LINT_LOG}" >> $GITHUB_ENV
+          echo "${EOF}" >> $GITHUB_ENV
+          echo "MD_LINT_STATUS=${MD_LINT_STATUS}" >> $GITHUB_ENV
+
+          echo "Linting front-matter"
+          FM_LINT_STATUS=0
+          FM_LINT_LOG=$(node scripts/front-matter_linter.js --fix true ${files_to_lint} 2>&1) || FM_LINT_STATUS=1
+          echo "FM_LINT_LOG<<${EOF}" >> $GITHUB_ENV
+          echo "${FM_LINT_LOG}" >> $GITHUB_ENV
+          echo "EOF" >> $GITHUB_ENV
```
Comment (OnkarRuikar) [inline .github/workflows/pr-review-lint.yml:66]:
```suggestion
          echo "${EOF}" >> $GITHUB_ENV
```

File: .github/workflows/pr-review-lint.yml (hunk 1 above)
Comment (OnkarRuikar) [reply .github/workflows/pr-review-lint.yml]: Done!

**Comment by yin1999** [comment]
Hi @OnkarRuikar, should we make the "add lint reviews to PRs" failed when there does have some formats need to be done (maybe we should rename this workflow). Just [as what I said in another PR](https://github.com/mdn/content/pull/28254#issuecomment-1658627135), this could help the maintainer to check if it's ready to merge PRs (if the contributer just mark the suggestions as resolved, the maintainer may not realize that the PR is not ready to be merged yet.

**Comment by OnkarRuikar** [comment]
> if the contributer just mark the suggestions as resolved, the maintainer may not realize that the PR is not ready to be merged yet

Good point! Now the job fails when there are any suggestions by the linters or raised issues.

**Comment by OnkarRuikar** [comment]
We are [not able to make it work for the PRs submitted from forks of guest users](https://github.com/reviewdog/reviewdog#graceful-degradation-for-pull-requests-from-forked-repositories). The [initial PR](https://github.com/mdn/content/pull/28228) showed those review comments because the PR author has direct push permissions to the repo so [their GITHUB_TOKEN in the workflow runs got write permissions](https://github.com/mdn/content/actions/runs/5698745354/job/15446945539#step:1:26). But PRs from guest users are [not getting the write permissions](https://github.com/mdn/content/actions/runs/5713300190/job/15478455678?pr=28295#step:1:26). We've tried to use a repo wide secret personal access token with PR write permissions but it is not visible in PRs from fork.

**Comment by yin1999** [comment]
> Hi @OnkarRuikar, is it possible that we could run the reviewdog in another workflow which could be triggered by "Add lint reviews to PRs" workflow. So we could use mdn-bot token (or other token which has write permission, maybe in this way, the GITHUB_TOKEN does have write permission) to report suggestions (This should keep the repository secure). But the reviewdog seems to only works with pull request event.
> 
> So I just check how the reviewdog collects enough info before submitting suggestions. The entry should be [here](https://github.com/reviewdog/reviewdog/blob/439aec2bbeb833f510426822a9db01e7edbf910a/cienv/cienv.go#L38-L40). And all the information seems to be provided by the event file, and it's location is provided by `GITHUB_EVENT_PATH` environment, see: https://github.com/reviewdog/reviewdog/blob/439aec2bbeb833f510426822a9db01e7edbf910a/cienv/github_actions.go#L69-L74, and https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables
> 
> So, I'm wondering is it possible that we could upload the event file and git diff content, and in another PR, we just download those files, and set up the `GITHUB_EVENT_PATH` to simulate the pull request event.

I just setup the [`pr-review-lint-companion`](https://github.com/yin1999/translated-content/blob/d115bef953c8bdc10252aaf1dd90f691eca6d0cd/.github/workflows/pr-review-lint-companion.yml) workflow, add modified the pr review lint workflow (yin1999/translated-content#17), the reviewdog just submmitted suggestions successfully (check https://github.com/yin1999/translated-content/pull/17#discussion_r1281285404) :)

/cc @LeoMcA I'm not sure if it's secure enough for us, do you have any thought?

**Comment by yin1999** [comment]
Hi @OnkarRuikar, one more question, I'm not really familiar with markdown-lint and front-matter-lint, do I just need to redirect stderr to a log file? or we need to combine stdout with stderr, and then redirect to a log file?

**Comment by OnkarRuikar** [comment]
> I think [pull_request_target](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) is the way forward here...

I'll try this approach in a new draft PR. This one has a lot of commits and history accrued.

**Comment by Josh-Cena** [comment]
Looks like a duplicate of #28363

**Pull request closed without merging**